### PR TITLE
Add fp16 support for Bert QA inference

### PR DESCRIPTION
**Pull request by MoisesHer**
## Description ##
Adding FP16 in QA-BERT inference

### Essentials ###
- [x] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [x] Code is well-documented

### Changes ###
- [x] Added option to use fp16 data type in run_squad script for evaluation (--eval_dtype). Will cast the network.
- [x] Propagated the datatype to masked_softmax, where was using fp32 always and causing nans if fp16 was selected

## Comments ##

cc @dmlc/gluon-nlp-team, @sxjscience

### CODE REVIEW

File: scripts/question_answering/run_squad.py [hunk 1]
```diff
@@ -311,13 +314,14 @@ def get_network(model_name,
     # Create the network
     Model, cfg, tokenizer, download_params_path, _ = \
         get_backbone(model_name, load_backbone=not backbone_path)
-    backbone = Model.from_cfg(cfg, use_pooler=False)
+    backbone = Model.from_cfg(cfg, use_pooler=False, dtype=dtype)
```
Comment (sxjscience) [inline scripts/question_answering/run_squad.py:317]: Would you try to use:
```python
cfg.defrost()
cfg.MODEL.dtype = dtype
cfg.freeze()

backbone = Model.from_cfg(cfg, use_pooler=False)
```
So there is no need to revise the cfg parsing logic.

File: scripts/question_answering/run_squad.py [hunk 2]
```diff
@@ -777,7 +784,9 @@ def predict_extended(original_feature,
 def evaluate(args, last=True):
     ctx_l = parse_ctx(args.gpus)
     cfg, tokenizer, qa_net = get_network(
-        args.model_name, ctx_l, args.classifier_dropout)
+        args.model_name, ctx_l, args.classifier_dropout, dtype=args.eval_dtype)
+    if args.eval_dtype == 'float16':
+        qa_net.cast('float16')
```
Comment (sxjscience) [inline scripts/question_answering/run_squad.py:789]: Is it possible to merge this into `get_network`? We may add tests in a later PR.

File: scripts/question_answering/run_squad.py [hunk 3]
```diff
@@ -311,13 +314,17 @@ def get_network(model_name,
     # Create the network
     Model, cfg, tokenizer, download_params_path, _ = \
         get_backbone(model_name, load_backbone=not backbone_path)
+    cfg.defrost()
+    cfg.MODEL.dtype = dtype
+    cfg.freeze()
```
Comment (szha) [inline scripts/question_answering/run_squad.py:319]: changing it here doesn't seem like a good general approach. would it be better to add the datatype support to `get_backbone`?

Comment (sxjscience) [reply scripts/question_answering/run_squad.py:319]: After a second thought. May be it's better to add dtype to `get_backbone`.

File: scripts/question_answering/run_squad.py (hunk 2 above)
Comment (MoisesHer) [reply scripts/question_answering/run_squad.py:789]: in the case of training, I think we do not want to cast if eval_type is fp16, that is the reason to do it here.
In the coming days I think we can AMP for training as well

File: scripts/question_answering/run_squad.py (hunk 3 above)
Comment (MoisesHer) [reply scripts/question_answering/run_squad.py]: ok, np, I made the change back

**Comment by sxjscience** [comment]
Will need to add test later.

**Pull request merged**